### ISSUE TITLE
Consider alternative to loading ImageReader via reflection

### ISSUE THREAD

**Issue by sloshy** [sksamuel/scrimage#217]
Hi, I'm working on upgrading the version of scrimage used within the [47degrees/sbt-microsites](https://github.com/47degrees/sbt-microsites/) project from 2.x to the latest version 4.x. This is an SBT build plugin, and by default SBT uses a limited classloader for build plugins that does not allow me to load any `ImageReader` classes.

So when I call `ImmutableImage.loader.fromFile(file)` or any other loader method, it will try to enumerate all of the instances of `ImageReader` before it starts reading the image. This is defined here:

https://github.com/sksamuel/scrimage/blob/79a012a7eed9b9847737323e7006315c31aad9e2/scrimage-core/src/main/java/com/sksamuel/scrimage/nio/ImageReaders.java#L23

This means at runtime, because my classloader will not allow me to load any of these classes, it fails when trying to read any image. I can work around the issue by manually setting the classloader like so:

```scala
val currentCL = Thread.currentThread.getContextClassLoader()
val desiredCL = classOf[ImageReader].getClassLoader()

try {
  Thread.currentThread.setContextClassLoader(desiredCL)
  //Do operations with scrimage here
} finally {
  Thread.currentThread.setContextClassLoader(currentCL)
}
```

This is rather hacky and I'd much prefer a proper solution were built into the library. I'm thinking maybe one of the following could work:
* Use a static list of image readers by default that the user could maybe override or add to if they wanted to load other image types
* Allow providing a classloader to the library as a parameter so it can be passed through to the class-loading code above

**Comment by sksamuel** [comment]
Hi,

Let me ask a question to clarify your issue. If your classloaoder can't load ImageReader classes, how will you decode the image?

**Comment by sloshy** [comment]
With the workaround as posted above, it gets a classloader that is capable of loading the `ImageReader` subclasses. I'm actually kind of surprised this solution works, to be honest.

To clarify the context in which the issue appears, say I copy that exact snippet of `StreamSupport` code that gets a list of all `ImageReader` from the classloader, but with Scala syntax (so `classOf` instead of `.class`, that's the only difference). If I run it in the SBT plugin code, it returns an empty list. If I do it in a normal application, or a test file, it works properly and returns a list of at least a few `ImageReader` instances.

So I am able to work around the issue but, from what I understand, libraries that do this kind of reflection should probably at least accept a classloader as input so it can be overridden in these cases, or they could maybe find some alternative to reflection that allows them to work in more esoteric use-cases like mine.

**Comment by sksamuel** [comment]
I understand then. Because SBT is not using the regular class loader and you need to use this functionality as part of the SBT build.

I think accepting a classloader is the best solution. I'll add overrides for the loader methods to do this and release 4.1.0

**Comment by sksamuel** [comment]
I was able to do this in a backwards compatible way.

`ImmutableImageLoader` now has a `withClassloader` method, and if set, will use that classloader to resolve image readers.
I've released 4.0.17

**Comment by sloshy** [comment]
Fantastic, thank you very much! I will use this going forward for my needs. Thank for for updating me.
